File: workspace/src/my_py_pkg/my_py_pkg/minimal_subscriber.py
import rclpy
from rclpy.executors import ExternalShutdownException
from rclpy.node import Node
from std_msgs.msg import String

class MinimalSubscriber(Node):
    """Subscriber example that prints messages to the console"""

    def __init__(self):
        """Constructor"""

        # Call the Node class constructor with the node name
        super().__init__('minimal_subscriber')

        # Create a subscription object
        self._subscription = self.create_subscription(
            String,
            'my_topic',
            self._listener_callback,
            10
        )

    def _listener_callback(self, msg):
        """Prints message to the console"""
        self.get_logger().info(f"Received message: {msg.data}")

def main(args=None):
    """Main entrypoint"""

    # Initialize and run node
    try:
        rclpy.init()
        node = MinimalSubscriber()
        rclpy.spin(node)

    # Catch ctrl+c or shutdown request
    except (KeyboardInterrupt, ExternalShutdownException):
        pass

    # Destroy node (now) and gracefully exit
    finally:
        if node is not None:
            node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()

if __name__ == '__main__':
    main()

File: workspace/src/my_cpp_pkg/src/minimal_subscriber.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/string.hpp"

/**
 * Subscriber example that prints messages to the console
 */
class MinimalSubscriber : public rclcpp::Node
{

public:

  /**
   * Constructor (call Node class constructor with node name)
   */
  MinimalSubscriber() : Node("minimal_subscriber")
  {

    // Create a subscription object
    subscription_ = this->create_subscription<example_interfaces::msg::String>(
      "my_topic", 
      10, 
      std::bind(
        &MinimalSubscriber::listener_callback, 
        this, 
        std::placeholders::_1));
  }

private:

  /**
   * Prints message to the console
   */
  void listener_callback(const example_interfaces::msg::String & msg)
  {
    RCLCPP_INFO(this->get_logger(), "Received message: '%s'", msg.data.c_str());
  }

  // Declare member variaables
  rclcpp::Subscription<example_interfaces::msg::String>::SharedPtr 
    subscription_;
};

/**
 * Main entrypoint
 */
int main(int argc, char * argv[])
{
  // Initialize ROS 2
  rclcpp::init(argc, argv);

  // Initialize and run node
  auto node = std::make_shared<MinimalSubscriber>();
  rclcpp::spin(node);

  // Cleanup
  rclcpp::shutdown();

  return 0;
}
